chore(backend): Use Go Mod instead of Bazel

dtzar opened the pull request:
**Description of your changes:**


**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

rmgogogo commented on backend/src/apiserver/go.mod:
```diff
@@ -0,0 +1,87 @@
+module apiserver
```
another name like "github.com/kubeflow/pipelines/apiserver"?

rmgogogo commented:
@Bobgy for a review.

Bobgy commented:
@dtzar Can you share why you made the PR?
There's already a `go.mod` in KFP repo's root folder, why do we need to create another one in a subfolder?

Bobgy commented:
And using go mod instead of bazel has already been merged in https://github.com/kubeflow/pipelines/pull/4373.

dtzar commented:
@Bobgy I would expect to be able to go to the part of kubeflow I want to build and simply type `go build` and to restore dependencies go to the directory where the go.mod exists and type `go get`.  Since there are no actual go files in the root directory, this doesn't work.  Sure, I could manually tell go to download from that root go.mod file, but then you lose go's native ability to auto-generate the dependencies it is using in the files themselves.

Also the restore commands in the `go.mod` are there to assist with local development.  I need to comment those out for final PR and update developer docs to let people know.

I haven't polished this PR (hence the draft), but just wanted to put it up there to get feedback on the rough direction before spending more time on it.

dtzar commented:
In general, I was hoping for [something like this](https://github.com/golang/go/wiki/Modules#faqs--multi-module-repositories).  There are still challenges I'm having with this PR as it stands.  

If I `cd` to folder `backend\src\apiserver` and then `go get` I get the following error even if I remove the root `kubeflow/pipelines` from the go.mod and delete this package it automatically shows up again:

```shell
...
go get .: ambiguous import: found package github.com/kubeflow/pipelines/backend/src/apiserver in multiple modules:
        github.com/kubeflow/pipelines/backend/src/apiserver (C:\src\go\src\github.com\kubeflow\pipelines\backend\src\apiserver)
        github.com/kubeflow/pipelines v0.0.0-20200910032744-41dc3865eb05 (C:\src\go\pkg\mod\github.com\kubeflow\pipelines@v0.0.0-20200910032744-41dc3865eb05\backend\src\apiserver)
```

Separate tangent issue for local dev which needs more thought/work to resolve.  
I uncomment/enable the `replace github.com/kubeflow/pipelines => ..\..\..` then I get:

```shell
...
client_manager.go:24:2: module github.com/kubeflow/pipelines@latest found (v0.0.0-20200910032744-41dc3865eb05, replaced by ..\..\..), but does not contain package github.com/kubeflow/pipelines/backend/src/common
client_manager.go:25:2: module github.com/kubeflow/pipelines@latest found (v0.0.0-20200910032744-41dc3865eb05, replaced by ..\..\..), but does not contain package github.com/kubeflow/pipelines/backend/src/model
main.go:32:2: module github.com/kubeflow/pipelines@latest found (v0.0.0-20200910032744-41dc3865eb05, replaced by ..\..\..), but does not contain package github.com/kubeflow/pipelines/backend/src/resnd/src/clientource                                                                                                                                                                                                 nd/src/common
main.go:33:2: module github.com/kubeflow/pipelines@latest found (v0.0.0-20200910032744-41dc3865eb05, replaced by ..\..\..), but does not contain package github.com/kubeflow/pipelines/backend/src/sernd/src/modelver                                                                                                                                                                                                   ource
client_manager.go:26:2: module github.com/kubeflow/pipelines@latest found (v0.0.0-20200910032744-41dc3865eb05, replaced by ..\..\..), but does not contain package github.com/kubeflow/pipelines/backevernd/src/storage  
```

If I do nothing from master today and go to `backend\src\apiserver` `go get` and `go build` will work, but then you aren't doing any versioning of the dependencies it pulls down and has no correlation to `go.mod` at root.

Also, unless I'm missing something there still is NO documentation on the proper way to build/run the go code locally.  Even when I build the API server and then run it local with `go run`, I have to pass some arguments which I need to figure out by parsing through the `Dockerfile`.

Bobgy commented:
I believe subfolders will find the root go.mod, see https://github.com/golang/go/wiki/Modules#should-i-have-multiple-modules-in-a-single-repository

The easiest setup is all in one module.

Bobgy commented:
We have settled on the new build without bazel, I believe we no longer need to keep this open
